app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala

```diff
@@ -19,51 +19,83 @@ package uk.gov.hmrc.securemessage.services
 import cats.data.NonEmptyList
 import com.google.inject.Inject
 import org.joda.time.DateTime
-import play.api.mvc.Result
-import play.api.mvc.Results.{ BadRequest, Conflict, Created }
 import uk.gov.hmrc.auth.core.{ AuthorisationException, Enrolments }
 import uk.gov.hmrc.http.HeaderCarrier
-import uk.gov.hmrc.securemessage.connectors.EmailConnector
+import uk.gov.hmrc.securemessage.connectors.{ ChannelPreferencesConnector, EmailConnector }
 import uk.gov.hmrc.securemessage.controllers.models.generic._
 import uk.gov.hmrc.securemessage.models.EmailRequest
 import uk.gov.hmrc.securemessage.models.core._
+import uk.gov.hmrc.securemessage.models.core.ParticipantType.{ Customer => PCustomer, System => PSystem }
 import uk.gov.hmrc.securemessage.repository.ConversationRepository
+import uk.gov.hmrc.securemessage.services.exception.SecureMessageException
+import cats.data._
+import cats.implicits._
+import uk.gov.hmrc.emailaddress.EmailAddress
 
 import scala.concurrent.{ ExecutionContext, Future }
-/*
-TODO: refactor so that service has no play dependencies, only core classes.
- */
 
-@SuppressWarnings(Array("org.wartremover.warts.ImplicitParameter"))
-class SecureMessageService @Inject()(repo: ConversationRepository, emailConnector: EmailConnector) {
+class SecureMessageService @Inject()(
+  repo: ConversationRepository,
+  emailConnector: EmailConnector,
+  channelPrefConnector: ChannelPreferencesConnector) {
+  @SuppressWarnings(Array("org.wartremover.warts.ImplicitParameter"))
+  def createConversation(conversation: Conversation)(
+    implicit hc: HeaderCarrier,
+    ec: ExecutionContext): Future[Either[SecureMessageException, Int]] =
+    (for {
+      cnv <- EitherT.right[SecureMessageException](getConversationWithEmailFromCds(conversation))
+      _   <- EitherT(repo.insertIfUnique(cnv))
+      res <- emailParticipants(cnv)
+    } yield res).value
 
-  def createConversation(conversationRequest: ConversationRequest, client: String, conversationId: String)(
+  @SuppressWarnings(Array("org.wartremover.warts.Equals"))
+  val hasEmail: Participant => Boolean = p =>
+    p.participantType == PSystem || (p.participantType == PCustomer && p.email.isDefined)
+
+  @SuppressWarnings(
+    Array(
+      "org.wartremover.warts.Nothing",
+      "org.wartremover.warts.Any",
+      "org.wartremover.warts.Equals",
+      "org.wartremover.warts.ImplicitParameter"))
+  private def emailParticipants(cnv: Conversation)(
```

Comment:
so many exceptions, can we change the code so there are no warts ?

Reply:
We pass implicts everywhere what is wrong with ==